pkg/gce-pd-csi-driver/controller.go

```diff
@@ -458,3 +496,52 @@ func diskIsAttachedAndCompatible(volume *compute.Disk, instance *compute.Instanc
 	}
 	return false, nil
 }
+
+func pickTopology(top *csi.TopologyRequirement) (string, error) {
+	reqTop := top.GetRequisite()
+	prefTop := top.GetPreferred()
+
+	// Pick the preferred toplogy in order
+	if len(prefTop) != 0 {
+		if prefTop[0].GetSegments() == nil {
+			return "", fmt.Errorf("preferred toplogies specified specified but no segments")
+		}
+
+		// GCE PD cloud provider Create has no restrictions so just create in top preferred zone
+		zone, err := getZoneFromSegment(prefTop[0].GetSegments())
+		if err != nil {
+			return "", fmt.Errorf("could not get zone from preferred topology: %v", err)
+		}
+		return zone, nil
+	} else if len(reqTop) != 0 {
+		r := rand.Intn(len(reqTop))
```

Comment:
The in-tree plugin has a round-robin mechanism to choose zones ([here](https://github.com/kubernetes/kubernetes/blob/845a55dbbda9a35cfd3659f1fe880d35a03dc44b/pkg/volume/util/util.go#L439)). Could we do something similar?

Reply:
we are not guaranteed to have a monotonically increasing id/name for the CSI Driver. Using a hash strategy could have unwanted behavior. We would have to actually keep track of which we've done or not to get round-robin going. I think a random pick is actually our best option her e

Reply:
From a backwards compatibility standpoint for in-tree migration, do we need to support this though?  cc @saad-ali

Reply:
This does not affect migration. There are two potential cases for the migration path:
1) The whole PD lifecycle is migrated to CSI, then CreateVolume is called and we pick a zone in whatever method we want (random here), all future references to this volume just continue to use the zone picked
2) PD is precreated in some zone, this logic is never used, we just use whatever zone was used in the creation (the round robin method)

I don't think theres any point after creation where we make assumptions on "how the zone was picked". so therefore it can't create any issues.

Reply:
If a user created a statefulset:
1. Before the migration, the PDs would be evenly spread across zones
2. After the migration, the PDs would be randomly spread across zones

The question is, do we want to retain the old spreading behavior for backwards compatibility?  Or should we just declare multi-zone in-tree provisioning broken and require users to used delayed provisioning going forward?